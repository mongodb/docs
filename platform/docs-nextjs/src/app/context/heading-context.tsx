"use client"

import { createContext, ReactNode, useContext } from 'react';

interface HeadingContextType {
  lastHeading: string;
  ignoreNextHeading: boolean;
}

const defaultVal = {
  lastHeading: '',
  ignoreNextHeading: false,
};

const HeadingContext = createContext<HeadingContextType>(defaultVal);

type HeadingContextProviderProps = {
  children: ReactNode;
  heading?: string;
  ignoreNextHeading?: boolean;
};

/**
 * Context provider to help track of page headings until the consuming component.
 * Headings are pushed into a list, with the last being the nearest heading, upwards in the AST tree.
 * Designed to be called in the init, so each child node of sections
 * as consumers can access the section header.
 */
const HeadingContextProvider = ({
  children,
  heading,
  ignoreNextHeading = false,
}: HeadingContextProviderProps) => {
  const { lastHeading: prevHeading, ignoreNextHeading: skipHeading } =
    useHeadingContext();

  const newHeading = skipHeading || !heading ? prevHeading : heading;

  return (
    <HeadingContext.Provider
      value={{
        lastHeading: newHeading,
        ignoreNextHeading: ignoreNextHeading ?? false,
      }}
    >
      {children}
    </HeadingContext.Provider>
  );
};

const useHeadingContext = () => {
  return useContext(HeadingContext);
};

export { HeadingContextProvider, useHeadingContext };
