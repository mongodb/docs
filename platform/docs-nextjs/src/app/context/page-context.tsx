import { createContext, useContext } from 'react';
import type { Root } from '@/types/ast';

export type PageTemplateType =
  | 'blank'
  | 'drivers-index'
  | 'document'
  | 'errorpage'
  | 'feature-not-avail'
  | 'instruqt'
  | 'landing'
  | 'openapi'
  | 'changelog'
  | 'search'
  | 'guide'
  | 'product-landing';

export interface PageContextType {
  page: Root | null;
  template: PageTemplateType | null;
  slug: string;
  tabsMainColumn: boolean | null;
  options: Root['options'] | null;
}

export const PageContext = createContext<PageContextType>({
  page: null,
  template: null,
  slug: '',
  tabsMainColumn: null,
  options: null,
});

export const usePageContext = () => {
  return useContext(PageContext);
};
