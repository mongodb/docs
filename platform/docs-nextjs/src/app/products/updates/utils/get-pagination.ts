/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
