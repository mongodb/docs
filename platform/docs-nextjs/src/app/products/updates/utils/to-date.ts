export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
