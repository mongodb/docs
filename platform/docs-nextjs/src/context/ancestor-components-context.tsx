import { createContext, useContext, type ReactNode } from 'react';

interface AncestorComponentContextType {
  table: boolean;
  procedure: boolean;
}

const defaultVal = {
  table: false,
  procedure: false,
};

const AncestorComponentContext = createContext<AncestorComponentContextType>(defaultVal);

type AncestorComponentName = 'table' | 'procedure';

export type AncestorComponentContextProviderProps = {
  children: ReactNode;
  component: AncestorComponentName;
};

/**
 * Context provider to help track ancestors of components without ambiguous prop drilling.
 * If nested within another component that uses this provder, previous ancestors are persisted.
 */
const AncestorComponentContextProvider = ({ children, component }: AncestorComponentContextProviderProps) => {
  const prevAncestors = useAncestorComponentContext();
  const newAncestors = { ...prevAncestors };

  if (component) {
    newAncestors[component] = true;
  }

  return <AncestorComponentContext.Provider value={newAncestors}>{children}</AncestorComponentContext.Provider>;
};

/**
 * By default, all ancestors will be false.
 */
const useAncestorComponentContext = () => {
  return useContext(AncestorComponentContext);
};

export { AncestorComponentContextProvider, useAncestorComponentContext };
