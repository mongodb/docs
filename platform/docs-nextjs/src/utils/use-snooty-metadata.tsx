'use client';

import { useContext, createContext } from 'react';
import type { RemoteMetadata } from '@/types/data';

const MetadataContext = createContext<RemoteMetadata | undefined>(undefined);

export const useSnootyMetadata = () => {
  const context = useContext(MetadataContext);

  if (!context) throw new Error('useSnootyMetadata must be used within a MetadataProvider');

  return context;
};

export const MetadataProvider = MetadataContext.Provider;
