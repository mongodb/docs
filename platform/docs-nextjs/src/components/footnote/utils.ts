import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
