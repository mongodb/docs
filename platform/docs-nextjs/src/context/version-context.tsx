import {
  createContext,
  useReducer,
  useEffect,
  useCallback,
  useContext,
  useRef,
  useMemo,
  type ReactNode,
  type Dispatch,
} from 'react';
import { useRouter } from 'next/navigation';
import { useSnootyMetadata } from '@/utils/use-snooty-metadata';
import { getLocalValue, setLocalValue } from '@/utils/browser-storage';
import type { Environments } from '@/utils/env-config';
import type { BranchData, Docset, Group } from '@/types/data';
import { getUrl } from '@/utils/url-utils';

type ActiveVersions = Record<string, string>;
type AvailableVersions = Record<string, BranchData[]>;
type AvailableGroups = Record<string, Group[]>;

// <-------------- begin helper functions -------------->
const STORAGE_KEY = 'activeVersions';
const LEGACY_GIT_BRANCH = 'legacy';

const getInitBranchName = (branches: BranchData[]) => {
  // Find 'current' branch as first option
  const currentBranch = branches.find(
    (b) => b.urlSlug === 'current' || b.gitBranchName === 'current' || b.urlAliases?.includes('current'),
  );
  if (currentBranch) {
    return currentBranch.gitBranchName;
  }
  const activeBranch = branches.find((b) => b.active);
  if (activeBranch) {
    return activeBranch.gitBranchName;
  }
  return branches[0]?.gitBranchName || null;
};

const findBranchByGit = (gitBranchName: string, branches?: BranchData[]) => {
  if (!branches || !branches.length) {
    return;
  }
  return branches.find(
    (b) => b.urlSlug === gitBranchName || b.gitBranchName === gitBranchName || b?.urlAliases?.includes(gitBranchName),
  );
};

// version state reducer helper fn
// overwrite current state with any new state attributes
const versionStateReducer = (state: ActiveVersions, newState: Partial<ActiveVersions>) => {
  // Cleans any undefined values - type safety
  const cleaned = Object.fromEntries(
    Object.entries(newState).filter((entry): entry is [string, string] => typeof entry[1] === 'string'),
  );

  return {
    ...state,
    ...cleaned,
  };
};
/**
 * get active branches for current product (from site metadata)
 *
 * @returns versions{} <product_name: branch_object[]>
 */
const getBranches = (docsets: Docset[]) => {
  const versions = getDefaultVersionsUnified(docsets);
  const groups = getDefaultGroupsUnified(docsets);
  return { versions, groups };
};

const getDefaultVersionsUnified = (docsets: Docset[]) => {
  const versions: { [k: string]: Docset['branches'] } = {};

  for (const docset of docsets) {
    // Skips non-versioned sites
    if (!docset.branches || docset.branches.length <= 1) {
      continue;
    }
    versions[docset.project] = docset.branches.filter((version: { active: boolean }) => version.active === true);
  }

  return versions;
};

const getDefaultGroupsUnified = (docsets: Docset[]) => {
  const groups: AvailableGroups = {};

  for (const docset of docsets) {
    groups[docset.project] = docset?.groups || [];
  }

  return groups;
};

interface ActiveVersionState {
  project: string;
  branch: string;
}

const getDefaultActiveVersions = ({ project, branch }: ActiveVersionState) => {
  // for current metadata.project, should always default to metadata.branch
  return { [project]: branch };
};
// <-------------- end helper functions -------------->

type VersionContextType = {
  siteBasePrefix: string;
  activeVersions: ActiveVersions;
  // active version for each product is marked is {[product name]: active version} pair
  setActiveVersions: Dispatch<Partial<ActiveVersions>>;
  availableVersions: AvailableVersions;
  availableGroups: AvailableGroups;
  onVersionSelect: (targetProject: string, gitBranchName: string) => void;
};

const VersionContext = createContext<VersionContextType>({
  siteBasePrefix: '',
  activeVersions: {},
  // active version for each product is marked is {[product name]: active version} pair
  setActiveVersions: () => {},
  availableVersions: {},
  availableGroups: {},
  onVersionSelect: () => {},
});

export const useVersionContext = () => {
  const context = useContext(VersionContext);

  if (!context) throw new Error('useVersionContext must be used within a VersionContextProvider');

  return context;
};

interface VersionContextProviderProps {
  children: ReactNode;
  slug: string;
  docsets: Docset[];
  env: Environments;
}

export const VersionContextProvider = ({ docsets, slug, env, children }: VersionContextProviderProps) => {
  const router = useRouter();
  const mountRef = useRef(true);

  const { versions, groups } = getBranches(docsets);
  const { project, branch } = useSnootyMetadata();

  const { metadata, siteBasePrefix } = useMemo(() => {
    const repoBranches = docsets.find((docset) => docset.project === project);
    return {
      metadata: { project, branch },
      siteBasePrefix: repoBranches?.prefix?.[getRepoBranchesPrefixEnv(env)] ?? '',
    };
  }, [project, branch, docsets, env]);

  // TODO: Might need to update this once we use this branch on a stitched project (DOP-5243 dependent)
  // TODO check whats going on here for 404 pages
  // tracks active versions across app
  const [activeVersions, setActiveVersions] = useReducer<
    React.Reducer<ActiveVersions, Partial<ActiveVersions>>,
    ActiveVersionState
  >(versionStateReducer, metadata, getDefaultActiveVersions);
  // update local storage when active versions change
  useEffect(() => {
    const existing = getLocalValue(STORAGE_KEY);
    setLocalValue(STORAGE_KEY, { ...existing, ...activeVersions });
    return () => {
      mountRef.current = false;
    };
  }, [activeVersions]);

  // handler for selecting version on multiple dropdowns
  const onVersionSelect = useCallback(
    (targetProject: string, gitBranchName: string) => {
      const updatedVersion: Record<string, string> = {};
      updatedVersion[targetProject] = gitBranchName;
      setActiveVersions(updatedVersion);

      // navigate to new URL only if from same project
      if (targetProject !== metadata.project) {
        return;
      }

      const targetBranch = findBranchByGit(gitBranchName, versions[metadata.project]);
      if (!targetBranch && gitBranchName !== LEGACY_GIT_BRANCH) {
        console.error(`target branch not found for git branch <${gitBranchName}>`);
        return;
      }
      const target =
        gitBranchName === LEGACY_GIT_BRANCH
          ? gitBranchName
          : targetBranch?.urlSlug || targetBranch?.urlAliases?.[0] || targetBranch?.gitBranchName;

      const urlTarget = getUrl(target, metadata.project, siteBasePrefix, slug);

      router.push(urlTarget);
    },
    [versions, metadata, siteBasePrefix, slug],
  );

  return (
    <VersionContext.Provider
      value={{
        siteBasePrefix,
        activeVersions,
        setActiveVersions,
        availableVersions: versions,
        availableGroups: groups,
        onVersionSelect,
      }}
    >
      {children}
    </VersionContext.Provider>
  );
};

const getRepoBranchesPrefixEnv = (env: Environments) => {
  switch (env) {
    case 'dotcomprd':
      return 'dotcomprd';
    case 'production':
      return 'prd';
    case 'dev':
    case 'dotcomstg':
    default:
      return 'dotcomstg';
  }
};
