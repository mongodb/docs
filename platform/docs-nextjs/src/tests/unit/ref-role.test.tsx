import { render } from '@testing-library/react';
import { mockLocation } from '@/tests/utils/mock-location';
import type { RefRoleNode } from '@/types/ast';
import { useSnootyMetadata } from '@/utils/use-snooty-metadata';
import RefRole from '@/components/ref-role';
import { VersionContextProvider } from '@/context/version-context';
import { mockDocsets } from '@/tests/utils/mock-docsets';

// data for this component
import mockData from '@/tests/data/ref-role.test.json';
const typedMockData = mockData as RefRoleNode;

jest.mock('@/utils/use-snooty-metadata', () => ({
  useSnootyMetadata: jest.fn(),
}));

beforeAll(() => {
  mockLocation({ hash: '' });
  (useSnootyMetadata as jest.Mock).mockImplementation(() => ({ project: 'docs-node', branch: 'v4.9' }));
});

it('renders correctly with versioned path', () => {
  const tree = render(
    <VersionContextProvider docsets={mockDocsets} slug="reference/program/mongos" env="production">
      <RefRole nodeChildren={typedMockData.children} fileid={typedMockData.fileid} url={typedMockData.url} />
    </VersionContextProvider>,
  );
  expect(tree.asFragment()).toMatchSnapshot();
});
