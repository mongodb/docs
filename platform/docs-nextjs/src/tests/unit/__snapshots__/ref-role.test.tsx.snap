// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders correctly with versioned path 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  display: inline;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-decoration-color: transparent;
  cursor: pointer;
  font-size: inherit;
  line-height: inherit;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: none;
  border: none;
  padding: 0;
  font-size: 13px;
  line-height: 20px;
  color: #016BF8;
  font-weight: 400;
  display: inline;
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-0:hover,
.emotion-0[data-hover='true'],
.emotion-0:focus-visible,
.emotion-0[data-focus='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-transition: text-decoration 150ms ease-in-out;
  transition: text-decoration 150ms ease-in-out;
  text-underline-offset: 4px;
  text-decoration-thickness: 2px;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover,
.emotion-0[data-hover='true'] {
  text-decoration-color: #E8EDEB;
}

.emotion-0:focus-visible,
.emotion-0[data-focus='true'] {
  text-decoration-color: #016BF8;
}

.emotion-0>span>code,
.emotion-0>code {
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-1 {
  font-size: 13px;
  line-height: 20px;
  display: inline;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  border-radius: 3px;
  font-family: 'Source Code Pro',Menlo,monospace;
  line-height: 20px;
  background-color: #F9FBFA;
  border: 1px solid #E8EDEB;
  color: #1C2D38;
  white-space: nowrap;
  font-size: unset;
  display: inline;
  color: var(--font-color-primary);
  background: var(--background-color-secondary);
  word-wrap: break-word;
  white-space: unset;
}

.lg-ui-0000:hover>.emotion-1 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.lg-ui-0000:hover>.emotion-1 {
  box-shadow: 0 0 0 3px #E8EDEB;
  border: 1px solid #C1C7C6;
}

.lg-ui-0000:focus-visible>.emotion-1 {
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC;
  border: 1px solid #016BF8;
}

a .emotion-1 {
  color: inherit;
}

<a
    class="lg-ui-0001 emotion-0"
    href="https://mongodb.com/docs/languages/node/v4.9/reference/program/mongos/#std-program-mongos"
    target="_self"
  >
    <span>
      <code
        class="emotion-1"
      >
        mongos
      </code>
    </span>
  </a>
</DocumentFragment>
`;
